File: lsh/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import scala.util.Random
import scala.collection.mutable
import java.io._



val conf = new SparkConf().setAppName("task1")
val sc = new SparkContext(conf)



val inputFile = args(0)
val outputFile = args(1)



val data = sc.textFile(inputFile).filter(row => row != "user_id,business_id,stars").map(line => line.split(","))



val allUsers = data.map(_(0))
val allBusinesses = data.map(_(1))

val businessUsersRDD = data.map(x => (x(1), x(0))).groupByKey().mapValues(_.toSet)
val distinctUsers = data.map(_(0)).distinct().collect()
val userIndex = distinctUsers.zipWithIndex.toMap



def hashFunctions(numFuncs: Int, maxVal: Int): Seq[(Int, Int)] = {
  (0 until numFuncs).map(_ => (Random.nextInt(maxVal), Random.nextInt(maxVal)))
}



def createSignatureMatrix(
  businessUsers: Iterable[(String, Set[String])],
  numFuncs: Int,
  userIndex: Map[String, Int],
  bins: Int,
  prime: Int
): Map[String, Seq[Int]] = {
  val hashFuncs = hashFunctions(numFuncs, bins)
  businessUsers.map { case (business, users) =>
    val signature = hashFuncs.map { case (a, b) =>
      users.map(user => ((a * userIndex(user) + b) % prime) % bins).min
    }
    business -> signature
  }.toMap
}



def lsh(signatures: Map[String, Seq[Int]], bandSize: Int): Map[Seq[Int], Seq[String]] = {
  val lshBuckets = mutable.Map[Seq[Int], mutable.Buffer[String]]().withDefaultValue(mutable.Buffer())
  for ((business, signature) <- signatures) {
    for (i <- 0 until signature.length by bandSize) {
      val band = signature.slice(i, i + bandSize)
      lshBuckets(band) += business
    }
  }
  lshBuckets.map { case (k, v) => (k, v.toSeq) }.toMap
}



def candidatePairs(lshBuckets: Map[Seq[Int], Seq[String]]): Set[(String, String)] = {
  lshBuckets.values.flatMap { bucket =>
    if (bucket.length > 1) bucket.combinations(2).map { case Seq(a, b) => (a, b) }
    else Seq.empty
  }.toSet
}




def jaccardSimilarity(
  candidatePairs: Set[(String, String)],
  businessUsers: Map[String, Set[String]],
  threshold: Double
): Map[(String, String), Double] = {
  candidatePairs.collect {
    case (bus1, bus2) if bus1 != bus2 =>
      val users1 = businessUsers(bus1)
      val users2 = businessUsers(bus2)
      val intersection = users1.intersect(users2).size
      val union = users1.union(users2).size
      val jaccardSim = intersection.toDouble / union
      if (jaccardSim >= threshold) (bus1, bus2) -> jaccardSim
  }.toMap
}



val businessUsersDict = businessUsersRDD.collectAsMap().toMap
val numHashes = 60
val prime = 987654321
val hashFuncs = hashFunctions(numHashes, distinctUsers.length)
val signatureMatrix = createSignatureMatrix(businessUsersDict, numHashes, userIndex, distinctUsers.length, prime)
val bandSize = 2
val buckets = lsh(signatureMatrix, bandSize)
val pairs = candidatePairs(buckets)
val similarityThreshold = 0.5
val similarPairs = jaccardSimilarity(pairs, businessUsersDict, similarityThreshold)



val writer = new PrintWriter(new File(outputFile))
writer.write("business_id_1,business_id_2,similarity\n")
similarPairs.toSeq.sortBy(_._1).foreach { case ((bus1, bus2), sim) =>
  writer.write(s"$bus1,$bus2,$sim\n")
}
writer.close()

sc.stop()

File: Homework 3/task1.py
from pyspark import SparkContext
import sys
import time
import random
from itertools import combinations



def make_hash_functions(num_funcs, max_val):
    return [(random.randint(1, max_val), random.randint(1, max_val)) for _ in range(num_funcs)]


def create_signature_matrix(business_users, num_funcs, user_index, bins, prime, hash_funcs):
    signatures = {}
    for business, users in business_users:
        signature = [min(((a * user_index[user] + b) % prime) % bins for user in users)
                     for a, b in hash_funcs]
        signatures[business] = signature
    return signatures


def lsh(signatures, band_size):
    lsh_buckets = {}
    for business, signature in signatures.items():
        for i in range(0, len(signature), band_size):
            band = tuple(signature[i:i + band_size])
            lsh_buckets.setdefault(band, []).append(business)
    return lsh_buckets



def make_candidate_pairs(lsh_buckets):
    pairs = set()
    for bucket in lsh_buckets.values():
        if len(bucket) > 1:
            pairs.update(combinations(sorted(bucket), 2))
    return pairs




def jaccard_similarity(candidate_pairs, business_users, threshold):
    similar_pairs = {}
    for bus1, bus2 in candidate_pairs:
        if bus1 != bus2: # Latest change
            users1, users2 = business_users[bus1], business_users[bus2]
            jaccard_sim = len(users1 & users2) / len(users1 | users2)
            if jaccard_sim >= threshold:
                similar_pairs[(bus1, bus2)] = jaccard_sim
    return similar_pairs


def main(input_file, output_file):
    sc = SparkContext(appName="task1")
    start_time = time.time()

    data = sc.textFile(input_file).filter(lambda row: row != "user_id,business_id,stars").map(lambda x: x.split(","))

    all_users = data.map(lambda x: x[0])
    all_business = data.map(lambda x: x[1])

    business_users_rdd = data.map(lambda x: (x[1], x[0])).groupByKey().mapValues(set)

    distinct_users = data.map(lambda x: x[0]).distinct().collect()

    user_index = {user: idx for idx, user in enumerate(distinct_users)}

    business_users_dict = business_users_rdd.collectAsMap()

    num_hashes = 60
    len_users = len(user_index)
    prime = 9876543210

    hash_funcs = make_hash_functions(num_hashes, len_users)
    #print(hash_funcs)

    signature_matrix = create_signature_matrix(business_users_dict.items(), num_hashes, user_index, len_users, prime, hash_funcs)

    band_size = 2
    buckets = lsh(signature_matrix, band_size)

    candidate_pairs = make_candidate_pairs(buckets)

    similarity_threshold = 0.5
    similar_pairs = jaccard_similarity(candidate_pairs, business_users_dict, similarity_threshold)

    with open(output_file, "w") as f:
        f.write("business_id_1,business_id_2,similarity\n")
        for (bus1, bus2), sim in sorted(similar_pairs.items()):
            f.write(f"{bus1},{bus2},{sim}\n")



if __name__ == '__main__':
    input_file, output_file = sys.argv[1], sys.argv[2]

    main(input_file, output_file)

# /opt/spark/spark-3.1.2-bin-hadoop3.2/bin/spark-submit task1.py ../resource/asnlib/publicdata/yelp_train.csv task1.csv